# CPE15 - Week 1: Python Basics for Data Science

**Programming for Data Science | Professional Elective 1 | AY 2026-2027**

| Syllabus element | Alignment |
|---|---|
| Course outcome | CO1 |
| Course objective | COBJ1 |
| Assessment connection | Diagnostic assessment and Quiz 1 preparation |
| Sustainable Development Goals | 4, 9 |

This notebook is designed for explanation, live coding, guided practice, and independent follow-through. Run it from top to bottom in a fresh kernel so that every result can be reproduced.


## Lecture map

            1. Values, variables, and types
2. Collections: teach each structure separately
3. Comparisons, logical operators, and decisions
4. Loops and comprehensions
5. Functions, methods, lambda, `map`, and `filter`
6. Integrated case study: sensor-quality report

            **Live-teaching rhythm:** define the idea → predict the result → run a focused example → inspect the saved output → explain the evidence → complete the practice task.

## Learning outcomes

            By the end of the session, students should be able to:

            - select suitable scalar and collection data types for an engineering problem;
- use comparisons, logical operators, decisions, and loops to express rules;
- transform collections with comprehensions and small functions;
- use lambda expressions, map, filter, methods, and tuple unpacking appropriately;
- trace and debug a short Python program and explain its output.

            ## Lecture route

            1. Values, names, types, and conversions
2. Lists, tuples, dictionaries, and sets taught separately
3. Boolean logic and decision branches
4. Loops and comprehensions
5. Functions, methods, lambda, map, and filter
6. An integrated sensor-quality case study


In [ ]:
from pathlib import Path
import platform
import random
import sys

random.seed(15)
ARTIFACTS = Path("artifacts")
ARTIFACTS.mkdir(exist_ok=True)

print(f"Python: {sys.version.split()}")
print(f"Platform: {platform.system()}")
print(f"Artifacts folder: {ARTIFACTS.resolve()}")


> **Reproducibility habit:** a notebook is not finished merely because it ran once. It should run in order from a restarted kernel, use explicit inputs, avoid hidden state, and explain the meaning of its outputs.


## 1. Values, variables, and types

A **value** is a piece of data. A **type** tells Python what that value represents and which operations are valid. A **variable** is a name bound to a value with the assignment operator `=`.

| Type | Example | Meaning in a data-science task | Typical operation |
|---|---|---|---|
| `str` | `"CEN-01"` | identifier or category label | `.strip()`, `.lower()` |
| `int` | `120` | whole-number count | addition, comparison |
| `float` | `3.28` | measured or calculated quantity | arithmetic, rounding |
| `bool` | `True` | logical state | `and`, `or`, `not` |
| `NoneType` | `None` | value is absent or not yet known | `is None` check |

The expression on the **right** of `=` is evaluated first; the resulting value is then assigned to the name on the **left**. Use names that preserve meaning and units, such as `voltage_v` rather than `x`.


In [ ]:
device_id = "SLSU-CPE-015"   # str: a label, not a quantity
sample_count = 120            # int: a whole-number count
voltage_v = 3.28              # float: a measured quantity
is_calibrated = True          # bool: a two-state condition
last_error = None             # None: no recorded value

scalar_examples = [device_id, sample_count, voltage_v, is_calibrated, last_error]
[(value, type(value).__name__) for value in scalar_examples]


**Reading the output.** Each tuple contains the stored value followed by its type name. Notice that `True` is a Boolean rather than the string `"True"`, and `None` is different from `0`, `False`, and an empty string.

### Case 1B — converting external text into a number

CSV files, forms, and serial messages frequently supply numeric-looking data as text. `float(raw_temperature)` performs an explicit conversion. Once converted, the value can participate in numerical formulas.

**Syntax trace**

- `float("29.6")` returns the floating-point value `29.6`.
- `*`, `/`, and `+` perform arithmetic using normal precedence.
- `f"{temperature_c:.1f}"` formats the number with one digit after the decimal point.


In [ ]:
raw_temperature = "29.6"
temperature_c = float(raw_temperature)
temperature_f = temperature_c * 9 / 5 + 32

print("Original type:", type(raw_temperature).__name__)
print("Converted type:", type(temperature_c).__name__)
print(f"{temperature_c:.1f} deg C = {temperature_f:.1f} deg F")


**Expected result.** The original type is `str`, the converted type is `float`, and the converted temperature is `85.3 deg F`. In real extraction code, conversion belongs inside validation because text such as `"missing"` cannot be converted with `float()`.


### Science communication lens

- **Audience:** A classmate or technical reader who needs to understand the rule without mentally executing the code.
- **Lead with the meaning:** The example converts a text temperature into a numerical value and then applies a stated unit-conversion equation.
- **Show the evidence:** Show the relevant input, type, operation, returned value or selected branch, unit, and received/retained counts when data are filtered.
- **State the boundary:** Code output demonstrates how the stated rule behaves; it does not establish that the input data, threshold, or scientific model is correct.

An effective explanation follows **claim → evidence → reasoning → limitation**. Define technical terms when first used, retain units and denominators, distinguish observed output from interpretation, and use wording proportional to the strength of the evidence.

### Practice task — your turn

**Task.** Create four raw sensor values stored as text, convert them to appropriate Python types, and print each converted value beside its type and unit.

**Before running your solution.** Predict the most relevant result property: value, type, shape, retained row count, numerical range, visual pattern, map behavior, or artifact destination.

**After running your solution.** Compare the output with your prediction. Report the evidence with applicable units and counts, explain the reasoning, and state one assumption or limitation.

In [ ]:
# YOUR TURN — Values, variables, and types
# Complete the specific task in the Markdown cell above.
# Keep input values, intermediate results, and final evidence easy to inspect.

# Write your solution below.
temperature_c_raw = "34.0"
days_used_raw = "120"
is_used_raw = "False"
room_id = "CEN-102B"

temperature_c = float(temperature_c_raw)
days_used = int(days_used_raw)
is_used = bool(is_used_raw)

print("Temperature: ", temperature_c, type(temperature_c).__name__)
print("Days used: ", days_used, type(days_used).__name__)
print("Is used: ", is_used, type(is_used).__name__)
print("Room ID: ", room_id, type(room_id).__name__)

## 2. Collections: teach each structure separately

A collection stores multiple values, but the four built-in collection types express different rules.

| Question | List | Tuple | Dictionary | Set |
|---|---|---|---|---|
| Written with | `[...]` | `(...)` | `{key: value}` | `{value, ...}` |
| Ordered | Yes | Yes | Yes, by insertion | Do not depend on display order |
| Mutable | Yes | No | Yes | Yes |
| Allows duplicates | Yes | Yes | Keys: no; values: yes | No |
| Main access | integer index | integer index / unpacking | key | membership test |
| Best mental model | editable sequence | fixed record | labeled record | unique-value collection |

We will use one focused example for each structure before combining them.


### 2.1 List — an ordered sequence that can change

A list is appropriate when order matters and items may be added, removed, or replaced. List positions start at index `0`.

**Syntax used below**

- `[28.4, 28.9, ...]` creates a list.
- `readings[0]` retrieves the first item.
- `readings[1:3]` creates a slice from index 1 up to, but not including, index 3.
- `.append(29.0)` mutates the existing list by adding one item at the end.


In [ ]:
readings = [28.4, 28.9, 29.1, 28.7]

print("Complete list:", readings)
print("First reading:", readings[0])
print("Middle slice:", readings[1:3])

readings.append(29.0)
print("After append:", readings)
print("Number of readings:", len(readings))


**Output explanation.** The first value is `28.4` because index `0` means the first position. The slice contains `28.9` and `29.1`; its stop index is excluded. After `.append()`, the same list has five values.

**When not to use it.** A list is a poor choice when the number and meaning of positions must remain fixed, or when each field needs a descriptive label.


### 2.2 Tuple — an ordered fixed record

A tuple is useful for a value whose positions have an agreed meaning and should not be changed accidentally. Coordinates are a common example: `(latitude, longitude)`.

**Syntax used below**

- `(14.1135, 121.5569)` creates a two-item tuple.
- `latitude, longitude = station_coordinate` **unpacks** the two positions into two names.
- A tuple supports indexing, but not item assignment or `.append()`.


In [ ]:
station_coordinate = (14.1135, 121.5569)
latitude, longitude = station_coordinate

print("Coordinate tuple:", station_coordinate)
print("Latitude by index:", station_coordinate[0])
print("Latitude by unpacking:", latitude)
print("Longitude by unpacking:", longitude)
print("Tuple length:", len(station_coordinate))


**Output explanation.** Index `0` and the name `latitude` refer to the same first value. Unpacking documents the meaning of both positions. The tuple has length 2 and remains unchanged.

Attempting `station_coordinate[0] = 14.2` would raise `TypeError` because tuples are immutable. Immutability protects the record from accidental item replacement; it does not validate whether latitude and longitude are correct.


### 2.3 Dictionary — a labeled record

A dictionary maps a unique **key** to a **value**. It is usually clearer than a list or tuple when a record has several fields with distinct meanings.

**Syntax used below**

- `{"id": "CEN-01", ...}` creates key-value pairs.
- `station["location"]` performs direct key lookup and raises `KeyError` if the key is absent.
- `station.get("operator", "not assigned")` provides a safe default.
- `station["mean_c"] = ...` adds a new pair or replaces the value of an existing key.


In [ ]:
station = {
    "id": "CEN-01",
    "location": "Lucban",
    "readings_c": [28.4, 28.9, 29.1, 28.7],
}

print("Station ID:", station["id"])
print("Location:", station["location"])
print("Operator:", station.get("operator", "not assigned"))

station["mean_c"] = round(sum(station["readings_c"]) / len(station["readings_c"]), 2)
print("Updated record:", station)


**Output explanation.** The key `"location"` returns `"Lucban"`. Because `"operator"` does not exist, `.get()` returns the stated default instead of stopping the program. The final assignment adds the derived field `"mean_c"` to the same dictionary.

**Important distinction.** Dictionary keys are unique; assigning the same key again updates its value. Dictionary values may repeat and may themselves be lists, tuples, dictionaries, or sets.


### 2.4 Set — unique values and membership operations

A set removes duplicates and is optimized for questions such as “has this status been observed?” Set display order should not be used to communicate rank or sequence.

**Syntax used below**

- `{"online", "maintenance", "online"}` creates a set; the repeated value is retained only once.
- `"offline" in observed_statuses` returns a Boolean membership result.
- `|`, `&`, and `-` mean union, intersection, and difference.
- Use `set()` for an empty set; `{}` creates an empty dictionary.


In [ ]:
observed_statuses = {"online", "maintenance", "online"}
allowed_statuses = {"online", "maintenance", "offline"}
critical_statuses = {"maintenance", "offline"}

print("Unique observed statuses:", sorted(observed_statuses))
print("Unique count:", len(observed_statuses))
print("Was offline observed?", "offline" in observed_statuses)
print("Observed critical statuses:", sorted(observed_statuses & critical_statuses))
print("Allowed but not observed:", sorted(allowed_statuses - observed_statuses))


**Output explanation.** Although `"online"` appeared twice in the literal, the set contains two unique statuses. `offline` membership is `False`; the intersection contains only `maintenance`; and the difference identifies `offline` as allowed but not observed. `sorted()` is used only to make the displayed output deterministic.


### 2.5 Choosing and combining collections

Use the operations your problem requires to choose the structure:

- Need an editable time-ordered series? Use a **list**.
- Need a fixed `(latitude, longitude)` pair? Use a **tuple**.
- Need named fields for one station? Use a **dictionary**.
- Need unique categories or fast membership checks? Use a **set**.

The next record deliberately combines the four types. This is now an **integration example**, not the first introduction to the concepts.


In [ ]:
monitoring_station = {
    "id": "CEN-01",                              # str value
    "coordinate": (14.1135, 121.5569),            # tuple value
    "readings_c": [28.4, 28.9, 29.1, 28.7],      # list value
    "observed_statuses": {"online", "maintenance"},  # set value
}

print("Record type:", type(monitoring_station).__name__)
print("Coordinate type:", type(monitoring_station["coordinate"]).__name__)
print("Readings type:", type(monitoring_station["readings_c"]).__name__)
print("Statuses type:", type(monitoring_station["observed_statuses"]).__name__)


**Interpretation.** The outer dictionary supplies field labels. Each field then chooses a value type that matches its own rule. This nested design is common in JSON-like records, but a set would need conversion to a list before standard JSON serialization.


### Science communication lens

- **Audience:** A classmate or technical reader who needs to understand the rule without mentally executing the code.
- **Lead with the meaning:** List, tuple, dictionary, and set encode different structural rules, so the chosen collection should be justified by the operations the problem requires.
- **Show the evidence:** Show the relevant input, type, operation, returned value or selected branch, unit, and received/retained counts when data are filtered.
- **State the boundary:** Code output demonstrates how the stated rule behaves; it does not establish that the input data, threshold, or scientific model is correct.

An effective explanation follows **claim → evidence → reasoning → limitation**. Define technical terms when first used, retain units and denominators, distinguish observed output from interpretation, and use wording proportional to the strength of the evidence.

### Practice task — your turn

**Task.** Represent one monitoring station four ways: a list of readings, a coordinate tuple, a metadata dictionary, and a set of observed statuses; perform one operation that demonstrates why each structure fits its role.

**Before running your solution.** Predict the most relevant result property: value, type, shape, retained row count, numerical range, visual pattern, map behavior, or artifact destination.

**After running your solution.** Compare the output with your prediction. Report the evidence with applicable units and counts, explain the reasoning, and state one assumption or limitation.

In [ ]:
# YOUR TURN — Collections: teach each structure separately
# Complete the specific task in the Markdown cell above.
# Keep input values, intermediate results, and final evidence easy to inspect.

# Write your solution below.


# monitoring station records as array
station_readings = [28.4, 28.9, 29.1, 28.7]
print("All Readings: ", station_readings)
print("First Reading: ", station_readings[0])
print("Last Reading: ", station_readings[-1])

# monitoring station records as tuples
station_coordinates = (14.1135, 121.5569)
longitude, latitude = station_coordinates
print("\nCoordinates: ", station_coordinates)
print("Longitude: ", longitude)
print("Latitude: ", latitude)

# monitoring station records as sets
station_statuses =  {"online", "maintenance", "online", "offline", "online"}
print("\nStation statuses: ", station_statuses)

# monitoring station as dictionary
monitoring_station = {
    "id": "CEN-01",
    "coordinate": station_coordinates,
    "readings": station_readings,
    "observed_statuses": station_statuses,
}
print("Full Monitoring Station Info: ", monitoring_station)


## 3. Comparisons, logical operators, and decisions

A comparison asks one question and produces `True` or `False`.

| Operator | Question | Example |
|---|---|---|
| `==` | Are the values equal? | `status == "online"` |
| `!=` | Are the values different? | `status != "offline"` |
| `<`, `<=` | Is the left value smaller? | `battery_v < 3.2` |
| `>`, `>=` | Is the left value larger? | `temp_c >= 38` |
| `in` | Is an item present? | `status in allowed_statuses` |

Do not confuse assignment `=` with equality comparison `==`.


In [ ]:
temperature_c = 38.2
battery_v = 3.05
status = "online"

print("At least 38 C:", temperature_c >= 38)
print("Battery below 3.2 V:", battery_v < 3.2)
print("Status is online:", status == "online")
print("Known status:", status in {"online", "maintenance", "offline"})


### 3.1 Combining Boolean conditions

- `A and B` is true only when both conditions are true.
- `A or B` is true when at least one condition is true.
- `not A` reverses a Boolean result.

Parentheses make a policy easier to read. The variable names below preserve each intermediate decision so it can be inspected independently.


In [ ]:
temperature_c = 38.2
humidity_pct = 82
battery_v = 3.05

hot = temperature_c >= 38
humid = humidity_pct >= 75
low_battery = battery_v < 3.2

heat_alert = hot and humid
service_needed = heat_alert or low_battery
sensor_invalid = not (0 <= humidity_pct <= 100)

print("hot:", hot, "humid:", humid, "low_battery:", low_battery)
print("heat_alert:", heat_alert)
print("service_needed:", service_needed)
print("sensor_invalid:", sensor_invalid)


### 3.2 Selecting exactly one branch with `if` / `elif` / `else`

Python tests branches from top to bottom. The first true condition runs, and the remaining branches are skipped. Put invalid-data checks before normal classifications so impossible measurements are not mislabeled as ordinary alerts.


In [ ]:
temperature_c = 38.2
humidity_pct = 82

if not (0 <= humidity_pct <= 100):
    status = "invalid sensor value"
elif temperature_c >= 38 and humidity_pct >= 75:
    status = "heat alert"
elif temperature_c >= 35:
    status = "high temperature"
else:
    status = "normal"

print("Selected branch:", status)


**Trace.** The humidity is valid, so the first branch is skipped. Both parts of the heat-alert condition are true, so `status` becomes `"heat alert"`; Python does not continue to the later `elif` or `else`.

### 3.3 Boundary cases

If a policy starts at exactly 38 °C, use `>= 38`. Test values just below, exactly at, and just above every boundary.


In [ ]:
def alert_level(temp_c):
    if temp_c >= 38:
        return "alert"
    if temp_c >= 35:
        return "watch"
    return "normal"

boundary_tests = [34.9, 35.0, 37.9, 38.0, 38.1]
[(value, alert_level(value)) for value in boundary_tests]


**Expected classifications.** `34.9` is normal; `35.0` and `37.9` are watch; `38.0` and `38.1` are alert. The exact-boundary results prove whether `>=` matches the stated policy.


### Science communication lens

- **Audience:** A classmate or technical reader who needs to understand the rule without mentally executing the code.
- **Lead with the meaning:** The selected branch follows from explicit comparison boundaries and Boolean conditions that can be tested at, below, and above each threshold.
- **Show the evidence:** Show the relevant input, type, operation, returned value or selected branch, unit, and received/retained counts when data are filtered.
- **State the boundary:** Code output demonstrates how the stated rule behaves; it does not establish that the input data, threshold, or scientific model is correct.

An effective explanation follows **claim → evidence → reasoning → limitation**. Define technical terms when first used, retain units and denominators, distinguish observed output from interpretation, and use wording proportional to the strength of the evidence.

### Practice task — your turn

**Task.** Classify the boundary voltages 207, 208, 253, and 254 V using explicit `if`/`elif`/`else` conditions, then explain why each boundary enters its reported category.

**Before running your solution.** Predict the most relevant result property: value, type, shape, retained row count, numerical range, visual pattern, map behavior, or artifact destination.

**After running your solution.** Compare the output with your prediction. Report the evidence with applicable units and counts, explain the reasoning, and state one assumption or limitation.

In [7]:
# YOUR TURN — Comparisons, logical operators, and decisions
# Complete the specific task in the Markdown cell above.
# Keep input values, intermediate results, and final evidence easy to inspect.

# Write your solution below.

voltage_1 = 207
voltage_2 = 208
voltage_3 = 253
voltage_4 = 254

def get_alert_message(voltage: int) -> str:
    message = None

    if voltage < 220:
        message = "Voltage is too low"
    elif voltage == 220:
        message = "Voltage is correct"
    else:
        message = "Voltage too high"

    return message

def print_voltage_info(voltage: int):
    message = get_alert_message(voltage)

    print(f"Voltage: {voltage}V \t Message: {message}")

print_voltage_info(voltage_1)
print_voltage_info(voltage_2)
print_voltage_info(voltage_3)
print_voltage_info(voltage_4)

Voltage: 207V 	 Message: Voltage is too low
Voltage: 208V 	 Message: Voltage is too low
Voltage: 253V 	 Message: Voltage too high
Voltage: 254V 	 Message: Voltage too high


## 4. Loops and comprehensions

Repetition can be expressed in several ways. Choose based on the control you need.

| Construct | Best use | Stop condition |
|---|---|---|
| `for` loop | visit each item in a collection | collection is exhausted |
| `while` loop | repeat while a state remains true | Boolean condition becomes false or `break` runs |
| comprehension | one readable transformation or filter | source collection is exhausted |

### 4.1 `for` loop — explicit multi-step processing

The loop variable `count` receives one list item per iteration. `.append()` records each converted result.


In [ ]:
raw_adc = [510, 515, 498, 1023, 505]
volts = []

for count in raw_adc:
    voltage = count / 1023 * 5.0
    volts.append(round(voltage, 3))
    print(f"ADC {count:4d} -> {voltage:.3f} V")

print("Converted list:", volts)


**Process trace.** Five input counts produce five iterations and five output voltages. A regular loop is useful here because the intermediate `voltage` is named and each conversion is printed for inspection.

### 4.2 Comprehensions — compact transformation and filtering

Read `[expression for item in source]` as “produce this expression for every item.” Adding `if condition` retains only matching items.


In [ ]:
raw_adc = [510, 515, 498, 1023, 505]

rounded_volts = [round(count / 1023 * 5.0, 3) for count in raw_adc]
plausible_volts = [value for value in rounded_volts if value < 4.9]

print("Transformed:", rounded_volts)
print("Filtered:", plausible_volts)
print("Rejected count:", len(rounded_volts) - len(plausible_volts))


**Output explanation.** The transformation keeps the list length at five. The filter removes the 5.0 V value, so four plausible values remain and the rejected count is one. A comprehension is suitable because each rule fits in one readable expression.

### 4.3 `while` loop — condition-controlled repetition

A `while` loop must update the state used by its condition. Otherwise it may never terminate. `break` exits immediately when success occurs.


In [ ]:
attempts_remaining = 3
simulated_results = [False, False, True]
attempt_index = 0

while attempts_remaining > 0:
    passed = simulated_results[attempt_index]
    print(f"Attempt {attempt_index + 1}: passed={passed}")
    if passed:
        break
    attempts_remaining -= 1
    attempt_index += 1

print("Final result:", "passed" if passed else "failed")


**Trace.** Attempts 1 and 2 fail, which reduces the remaining count and advances the index. Attempt 3 succeeds, so `break` exits the loop. This pattern models retries; production code would also handle exceptions and real timing constraints.


### Science communication lens

- **Audience:** A classmate or technical reader who needs to understand the rule without mentally executing the code.
- **Lead with the meaning:** The loop or comprehension produces one traceable transformation or selection from the stated input collection.
- **Show the evidence:** Show the relevant input, type, operation, returned value or selected branch, unit, and received/retained counts when data are filtered.
- **State the boundary:** Code output demonstrates how the stated rule behaves; it does not establish that the input data, threshold, or scientific model is correct.

An effective explanation follows **claim → evidence → reasoning → limitation**. Define technical terms when first used, retain units and denominators, distinguish observed output from interpretation, and use wording proportional to the strength of the evidence.

### Practice task — your turn

**Task.** Use a loop to convert five Celsius readings to Fahrenheit, then use a comprehension to retain only readings at or above 30 °C and reconcile the received and retained counts.

**Before running your solution.** Predict the most relevant result property: value, type, shape, retained row count, numerical range, visual pattern, map behavior, or artifact destination.

**After running your solution.** Compare the output with your prediction. Report the evidence with applicable units and counts, explain the reasoning, and state one assumption or limitation.

In [8]:
# YOUR TURN — Loops and comprehensions
# Complete the specific task in the Markdown cell above.
# Keep input values, intermediate results, and final evidence easy to inspect.

# Write your solution below.

temps = [25, 93, 30.0, 22.9, 37.7]

print("Raw Readings: ", temps)

accepted_temps_f = [temp for temp in temps if temp <= 30]
rejected_count = len(temps) - len(accepted_temps_f)

# use loop to convert celcius to farenheit
for i in range(len(temps)):
    temps[i] = temps[i] * 9 / 5 + 32

measurements = {}
measurements["accepted farenheits"] = accepted_temps_f
measurements["farenheit"] = temps;

report = {
    "Recieved Temperatures" : len(temps),
    "Accepted Temperatures" : len(accepted_temps_f),
    "Rejected Temperatures" : rejected_count,
    "Measurements": measurements
}

report

Raw Readings:  [25, 93, 30.0, 22.9, 37.7]


{'Recieved Temperatures': 5,
 'Accepted Temperatures': 3,
 'Rejected Temperatures': 2,
 'Measurements': {'accepted farenheits': [25, 30.0, 22.9],
  'farenheit': [77.0, 199.4, 86.0, 73.22, 99.86]}}

## 5. Functions, methods, lambda, `map`, and `filter`

These five terms all involve behavior, but they solve different problems. Treating them as interchangeable hides how data move through a program.

| Construct | What it is | How it is called | Best use |
|---|---|---|---|
| function | named reusable operation | `function(argument)` | logic that deserves a name, tests, and documentation |
| method | operation supplied by an object's type | `object.method(argument)` | behavior naturally associated with that object |
| lambda | anonymous one-expression function | created with `lambda`, then called or passed | a short local key or predicate |
| `map` | applies one function to every item | `map(function, iterable)` | an existing function should transform every element |
| `filter` | keeps items accepted by a predicate | `filter(predicate, iterable)` | an existing predicate should select elements |

We will examine each construct independently, trace its input and return value, and state when another construct would be clearer.


### 5.1 Function — a named input-to-output rule

A function packages a reusable process. The definition below has four parts:

```python
def celsius_to_fahrenheit(temp_c: float) -> float:
    return temp_c * 9 / 5 + 32
```

- `def` begins a function definition.
- `celsius_to_fahrenheit` is the function name.
- `temp_c` is a **parameter**, the local name that will receive an input.
- `: float` and `-> float` are type hints for readers and tools; Python does not automatically enforce them.
- `return` sends the calculated value back to the caller and ends that function call.
- In `celsius_to_fahrenheit(29.0)`, `29.0` is the **argument** supplied to `temp_c`.

**When to use a regular function:** use one when a rule will be reused, tested, documented, validated, or changed independently. Prefer returning a value so another calculation, plot, table, or test can reuse it.


In [ ]:
def celsius_to_fahrenheit(temp_c: float) -> float:
    # Convert a Celsius temperature to Fahrenheit.
    return temp_c * 9 / 5 + 32

room_temperature_c = 29.0
room_temperature_f = celsius_to_fahrenheit(room_temperature_c)

print("Input argument:", room_temperature_c, "deg C")
print("Returned value:", room_temperature_f, "deg F")
print("Returned value reused:", round(room_temperature_f + 1, 1), "deg F")


**How it worked.** The call assigns `29.0` to the local parameter `temp_c`. Python evaluates the formula and returns `84.2`. Because the value is returned rather than only printed, the caller can store it, add to it, test it, or pass it to another function.

**When not to create a new function:** a one-time obvious expression may be clearer directly in the surrounding calculation. Do not create a function merely to hide an unexplained formula; its name, parameters, units, and assumptions must communicate the rule.

**Science-communication model.** Say: “Using the stated conversion equation, an input of 29.0 °C corresponds to 84.2 °F.” This identifies the method, input, unit, and result without claiming that the original sensor measurement was accurate.


### 5.2 Method — behavior attached to a particular object

A method is accessed with dot syntax: `object.method(arguments)`. The object's type determines which methods exist and what they return.

- String method: `raw_name.strip()` returns a new string with outer whitespace removed.
- List method: `samples.append(29.1)` changes the existing list and returns `None`.
- Dictionary method: `record.get("operator", "unassigned")` returns a value or a default.

The distinction between **returning a new value** and **mutating the object in place** is essential.

**When to use a method:** use it when the operation belongs naturally to the object's type—for example, cleaning a string, appending to a list, or retrieving a dictionary key safely. Check the documentation when you do not know whether the method mutates, returns, or does both.


In [ ]:
raw_station_name = "  Cen Laboratory  "
clean_station_name = raw_station_name.strip().upper()

samples = [28.4, 28.9]
append_return = samples.append(29.1)

station_record = {"id": "CEN-01", "status": "online"}
operator_name = station_record.get("operator", "unassigned")

print("Original string remains:", repr(raw_station_name))
print("New cleaned string:", clean_station_name)
print("List after append:", samples)
print("Return value of append:", append_return)
print("Dictionary default:", operator_name)


**How each method worked.** `.strip()` returned a new string, and `.upper()` operated on that returned string. `.append()` modified `samples` and returned `None`; therefore `samples = samples.append(29.1)` would incorrectly replace the list with `None`. `.get()` did not modify the dictionary; it returned the default because the requested key was absent.

**Science-communication model.** Explain the transformation and its boundary: “Station labels were standardized by removing outer whitespace and converting letters to uppercase; this improves label consistency but does not verify that the station name itself is correct.”


### 5.3 Lambda — a temporary one-expression function

A lambda expression has this pattern:

```python
lambda parameters: returned_expression
```

`lambda item: item[1]` receives one item and returns its second element. It has no statement block, no explicit `return`, and is normally used immediately as an argument such as a sorting key.

**When to use lambda:** use it for a short, obvious rule needed at one location—commonly `key=` in `sorted`, or a simple predicate passed to another function.

**When not to use lambda:** choose `def` if the rule needs a meaningful name, type hints, a docstring, validation, multiple expressions, reuse, or unit tests. A lambda should reduce clutter, not compress important scientific logic until it becomes difficult to audit.


In [11]:
devices = [
    ("D-03", 88.4),
    ("D-01", 97.2),
    ("D-02", 91.0),
]

quality_key = lambda item: item[1]
ranked_devices = sorted(devices, key=quality_key, reverse=True)

print("Key returned for D-03:", quality_key(devices[0]))
print("Ranked devices:", ranked_devices)


Key returned for D-03: 88.4
Ranked devices: [('D-01', 97.2), ('D-02', 91.0), ('D-03', 88.4)]


**How it worked.** `sorted` sends each device tuple to `quality_key`. The lambda returns `88.4`, `97.2`, or `91.0`; those returned scores control the ordering. `reverse=True` places the largest score first. The tuples themselves remain unchanged.

A named equivalent would be:

```python
def quality_score(device_record):
    return device_record[1]
```

The named version is preferable if the score location, missing-value behavior, or validation rule needs explanation.

**Science-communication model.** Say: “Devices were ranked by the supplied quality-score field, from highest to lowest.” Do not say “D-01 is the best device” unless the score definition, uncertainty, and decision criteria justify that broader claim.


### 5.4 `map` — apply one transformation to every item

`map(function, iterable)` takes one item at a time from the iterable, calls the function with that item, and yields each returned value.

```text
[27.5, 29.0, 31.2]
        │ send each value to celsius_to_fahrenheit
        ▼
[81.5, 84.2, 88.16]
```

In Python 3, `map` returns a lazy iterator. `list(...)` is used below to materialize all results for inspection.

**When to use `map`:** use it when a named function already expresses the transformation clearly and should be applied unchanged to every item. A list comprehension is often clearer when the transformation is short or needs to be read in place.


In [ ]:
temperatures_c = [27.5, 29.0, 31.2]

mapped_iterator = map(celsius_to_fahrenheit, temperatures_c)
print("Object returned by map:", type(mapped_iterator).__name__)

temperatures_f = list(mapped_iterator)
print("Mapped Fahrenheit values:", [round(value, 2) for value in temperatures_f])

comprehension_f = [celsius_to_fahrenheit(value) for value in temperatures_c]
print("Same as comprehension:", temperatures_f == comprehension_f)


**How it worked.** `map` did not call the conversion when the iterator was created. Converting it to a list requested each result: the function received 27.5, then 29.0, then 31.2, and returned one Fahrenheit value for each input. The output length remains three because `map` transforms rather than filters.

**Important behavior:** once a map iterator has been consumed, iterating over that same object again produces no values. Call `map(...)` again to create a fresh iterator.

**Science-communication model.** Report both the method and the population: “The Celsius-to-Fahrenheit function was applied to all three recorded temperatures; no readings were excluded.” This makes the transformation and denominator explicit.


### 5.5 `filter` — retain items that satisfy a predicate

`filter(predicate, iterable)` sends each item to a **predicate**, a function that returns `True` or `False`. Items producing `True` are yielded; items producing `False` are omitted.

```text
27.5 ── is_at_least_30? False ── reject
29.0 ── is_at_least_30? False ── reject
31.2 ── is_at_least_30? True  ── retain
```

**When to use `filter`:** use it when an existing named predicate clearly states the selection policy. A comprehension such as `[x for x in values if rule(x)]` is often easier to read when transformation and filtering are combined.

Filtering changes the denominator. A scientific report should preserve received, retained, and rejected counts and should state the rule that caused exclusion.


In [ ]:
def is_at_least_30(temp_c: float) -> bool:
    return temp_c >= 30

temperatures_c = [27.5, 29.0, 31.2]
filtered_iterator = filter(is_at_least_30, temperatures_c)
retained_temperatures = list(filtered_iterator)

rejected_count = len(temperatures_c) - len(retained_temperatures)
print("Received:", len(temperatures_c))
print("Retained:", retained_temperatures)
print("Rejected count:", rejected_count)

equivalent_filter = [value for value in temperatures_c if is_at_least_30(value)]
print("Same as comprehension:", retained_temperatures == equivalent_filter)


**How it worked.** The predicate returned `False` for 27.5 and 29.0 and `True` for 31.2, so one of three values was retained. The predicate classifies values; `filter` performs the iteration and selection.

**Do not communicate only the surviving value.** A transparent statement is: “One of three recorded temperatures met the illustrative threshold of at least 30 °C; two were excluded by that rule.” This reports the boundary and denominator. It does not imply that 30 °C is a scientifically or legally established danger threshold.


### 5.6 Decision guide and integrated callable pipeline

| If you need to... | Prefer | Why |
|---|---|---|
| name and test a reusable scientific rule | `def` function | the contract and assumptions remain visible |
| use behavior belonging to a value's type | method | the object supplies the operation |
| provide one tiny local key | lambda | avoids a distant one-use definition |
| apply an existing function to every item | `map` | separates iteration from transformation |
| retain items accepted by an existing predicate | `filter` | separates iteration from selection |
| transform or filter with a short visible expression | comprehension | often reads most directly in Python |

The following pipeline combines the constructs only after each one has been studied independently.


In [ ]:
raw_readings = [" 28.4 ", "missing", "31.2", " 29.7"]

def parse_temperature(text):
    cleaned = text.strip()                 # string method
    if cleaned.lower() == "missing":      # string method
        return None
    return float(cleaned)

def is_observed(value):
    return value is not None

parsed = list(map(parse_temperature, raw_readings))
observed = list(filter(is_observed, parsed))
ranked = sorted(observed, key=lambda value: value, reverse=True)

callable_report = {
    "received": len(raw_readings),
    "parsed_values": parsed,
    "observed_count": len(observed),
    "missing_count": len(raw_readings) - len(observed),
    "ranked_observed_c": ranked,
}
callable_report


**Trace of the integrated pipeline**

1. `map` sends each raw string to the named `parse_temperature` function.
2. String methods standardize each input inside that function.
3. The missing label becomes `None`; numerical text becomes `float`.
4. `filter` uses the named predicate `is_observed` to remove `None`.
5. `sorted` uses a simple lambda key and descending order.
6. The report preserves received, observed, and missing counts.

**Audience-ready explanation.** “Four text records were processed. Three contained numerical temperatures and one was labeled missing. The three observed values were ranked for display; the missing record was counted but excluded from numerical ordering.” This is concise, transparent about exclusion, and avoids presenting the ranked values as a complete or representative temperature study.


### Science communication lens

- **Audience:** A classmate or technical reader who needs to understand the rule without mentally executing the code.
- **Lead with the meaning:** Functions, methods, lambda, map, and filter differ in where behavior is defined, how input is supplied, what is returned, and whether values are transformed or removed.
- **Show the evidence:** Show the relevant input, type, operation, returned value or selected branch, unit, and received/retained counts when data are filtered.
- **State the boundary:** Code output demonstrates how the stated rule behaves; it does not establish that the input data, threshold, or scientific model is correct.

An effective explanation follows **claim → evidence → reasoning → limitation**. Define technical terms when first used, retain units and denominators, distinguish observed output from interpretation, and use wording proportional to the strength of the evidence.

### Practice task — your turn

**Task.** Define a named Celsius-to-Fahrenheit function, call one string method, use a lambda with `map` for conversion, and use `filter` for a stated temperature rule; compare what each callable receives and returns.

**Before running your solution.** Predict the most relevant result property: value, type, shape, retained row count, numerical range, visual pattern, map behavior, or artifact destination.

**After running your solution.** Compare the output with your prediction. Report the evidence with applicable units and counts, explain the reasoning, and state one assumption or limitation.

In [ ]:
# YOUR TURN — Functions, methods, lambda, map, and filter
# Complete the specific task in the Markdown cell above.
# Keep input values, intermediate results, and final evidence easy to inspect.
# Write your solution below.

# convert celcius to farenheit
def celc_to_farenheit (celcius):
    return celcius * 9 / 5 + 32

# accept temperatures less than or equal to 100 degree farenheit
validate_temp_f = lambda temp_f: temp_f <= 100

celcius_list = [36, 38, 31, 32, 40, 34]
farenheit_list = list(map(celc_to_farenheit, celcius_list))
accepted_temp_f = list(filter(validate_temp_f, farenheit_list))

print("farenheit: ")
print(farenheit_list, "\n")

print("accepted farenheits")
print(accepted_temp_f)

## 6. Integrated case study: sensor-quality report

Integration comes after the individual constructs are understood. This case uses:

1. a **list** of station-record **dictionaries** as input;
2. a function containing ordered validation branches;
3. a loop that adds one classification to each record;
4. a comprehension that selects usable temperatures; and
5. a summary dictionary that preserves received, usable, and rejected counts.

### Step 1 — declare records and define the validation policy

The function returns the first applicable label. Temperature validity is checked before the low-battery warning because an impossible temperature must not be reported as usable.


In [ ]:
sensor_rows = [
    {"station": "CEN-A", "temperature_c": 29.4, "battery_v": 3.91},
    {"station": "CEN-B", "temperature_c": 31.7, "battery_v": 3.42},
    {"station": "CEN-C", "temperature_c": 99.0, "battery_v": 3.88},
    {"station": "CEN-D", "temperature_c": 30.1, "battery_v": 2.95},
]

def classify_record(row):
    if not (-10 <= row["temperature_c"] <= 60):
        return "invalid temperature"
    if row["battery_v"] < 3.2:
        return "low battery"
    return "usable"


### Step 2 — classify every record

During each iteration, `row` refers to one dictionary from `sensor_rows`. Assigning `row["quality_status"]` mutates that dictionary by adding a new labeled field.


In [ ]:
for row in sensor_rows:
    row["quality_status"] = classify_record(row)

for row in sensor_rows:
    print(row["station"], "->", row["quality_status"])


**Expected labels.** CEN-A and CEN-B are usable, CEN-C has an invalid temperature, and CEN-D has a low battery. Printing one station per line makes the classification trace visible before aggregation.

### Step 3 — select valid evidence and create a reconciled report

A trustworthy report should make this identity checkable:

`records received = records usable + records rejected`


In [ ]:
usable_temps = [
    row["temperature_c"]
    for row in sensor_rows
    if row["quality_status"] == "usable"
]

records_received = len(sensor_rows)
records_usable = len(usable_temps)
records_rejected = records_received - records_usable

report = {
    "records_received": records_received,
    "records_usable": records_usable,
    "records_rejected": records_rejected,
    "mean_usable_temperature_c": round(
        sum(usable_temps) / records_usable, 2
    ),
    "status_counts": {
        status: sum(row["quality_status"] == status for row in sensor_rows)
        for status in {row["quality_status"] for row in sensor_rows}
    },
}

assert report["records_received"] == report["records_usable"] + report["records_rejected"]
report


**Interpretation.** Two of four records are usable, so the usable rate is 50%. The mean is calculated only from CEN-A and CEN-B; it must not be described as the mean of all received readings. The status counts explain why the other two records were excluded.


### Science communication lens

- **Audience:** A classmate or technical reader who needs to understand the rule without mentally executing the code.
- **Lead with the meaning:** The sensor report classifies four received records, retains two as usable, and preserves the two rejection reasons before calculating a mean.
- **Show the evidence:** Show the relevant input, type, operation, returned value or selected branch, unit, and received/retained counts when data are filtered.
- **State the boundary:** Code output demonstrates how the stated rule behaves; it does not establish that the input data, threshold, or scientific model is correct.

An effective explanation follows **claim → evidence → reasoning → limitation**. Define technical terms when first used, retain units and denominators, distinguish observed output from interpretation, and use wording proportional to the strength of the evidence.

### Practice task — your turn

**Task.** Build a six-record sensor-quality report that preserves rejection reasons, computes a mean only from accepted observations, and reports received, accepted, rejected, and missing counts.

**Before running your solution.** Predict the most relevant result property: value, type, shape, retained row count, numerical range, visual pattern, map behavior, or artifact destination.

**After running your solution.** Compare the output with your prediction. Report the evidence with applicable units and counts, explain the reasoning, and state one assumption or limitation.

In [ ]:
# YOUR TURN — Integrated case study: sensor-quality report
# Complete the specific task in the Markdown cell above.
# Keep input values, intermediate results, and final evidence easy to inspect.

# Write your solution below.

sensor_rows = [
    {"station": "CEN-A", "temperature_c": 29.4},
    {"station": "CEN-B", "temperature_c": None},
    {"station": "CEN-C", "temperature_c": 99.0},
    {"station": "CEN-D", "temperature_c": 32.5},
    {"station": "CEN-E", "temperature_c": 31.9},
    {"station": "CEN-F", "temperature_c": 35.5}
]

def validate_row(row):
    errors = []

    temp = row["temperature_c"]

    if temp is None:
        errors.append("Temperature not found")

    temperature_out_of_bound = temp is not None and (temp <= -10 or temp >= 40)
    if temperature_out_of_bound:
        errors.append("Temperature out of bound")

    row["errors"] = errors
    return row;

def get_mean(rows: list, row_name: str):
    measures = [row[row_name] for row in rows]
    return sum(measures) / len(rows)



sensor_rows = [validate_row(row) for row in sensor_rows]
accepted_rows = list( filter(lambda row: len(row["errors"]) <= 0, sensor_rows) )
rejected_count = len(sensor_rows) - len(accepted_rows)
accepted_mean = get_mean(accepted_rows, "temperature_c")
missing_count = len(accepted_rows) + rejected_count - len(sensor_rows)

assert len(accepted_rows) + rejected_count == len(sensor_rows)
report = {
    "Recieved Records" : len(sensor_rows),
    "Rejected Records" : rejected_count,
    "Accepted" : {
        "Records": len(accepted_rows),
        "Mean Temperature": accepted_mean
    },
    "Missing Records" : missing_count
}

report

## Common mistakes and debugging habits

| Mistake | Why it fails | Better check |
|---|---|---|
| using `=` inside a comparison | `=` performs assignment | use `==` to compare |
| assuming index 1 is the first item | sequence indexing starts at zero | inspect `items[0]` and `len(items)` |
| assigning the result of `.append()` | `.append()` returns `None` | call it on its own line |
| changing a tuple item | tuples are immutable | create a new tuple |
| treating set display order as meaningful | sets model uniqueness, not rank | use a list or sort only for display |
| allowing a `while` condition to remain unchanged | the loop may never terminate | trace the state update |
| printing inside every function | printed text is hard to reuse | return a value, then print at the boundary |

Read the final traceback line first, locate the failing line, inspect the relevant types and values, and reduce the problem to the smallest reproducible example.


In [ ]:
def safe_mean(values):
    # Return the arithmetic mean, or None for an empty collection.
    if not values:
        return None
    return sum(values) / len(values)

assert safe_mean([]) is None
assert safe_mean([2, 4, 6]) == 4
print("Boundary checks passed.")


The first assertion tests the empty-input branch; the second tests an ordinary three-value calculation. A passing assertion displays nothing, so the final message confirms that execution reached the end.

## Guided practice

Create a function `summarize_power` that accepts a list of `(voltage, current)` tuples. It should calculate power in watts for each pair, retain values from 0 to 100 W, and return the valid values, rejected count, and valid mean.

Before viewing the solution, identify where the design uses a tuple, list, comprehension, filter condition, function return value, and dictionary.


In [ ]:
def summarize_power(measurements):
    powers = [voltage * current for voltage, current in measurements]
    valid = [round(power, 2) for power in powers if 0 <= power <= 100]
    return {
        "valid_power_w": valid,
        "rejected": len(powers) - len(valid),
        "mean_valid_w": round(safe_mean(valid), 2) if valid else None,
    }

practice_measurements = [(12.0, 1.5), (5.0, 0.8), (24.0, 6.0)]
practice_report = summarize_power(practice_measurements)
assert practice_report["rejected"] == 1
practice_report


**Solution reading.** Each tuple is unpacked into `voltage` and `current`. The first comprehension produces all power values; the second retains values within the accepted range. The returned dictionary keeps the valid values, rejected count, and mean under explicit labels.


## Independent challenge

Build a quality-control report for at least eight environmental readings represented as dictionaries. Include one missing or invalid value, document the validation rule, classify each record, compute at least two summaries, and write a three-sentence interpretation.

**Submission expectation:** include readable code, meaningful variable names, a short interpretation, and evidence that the notebook was restarted and run from top to bottom.


## Key takeaways

- Types determine valid operations and should preserve data meaning.
- Lists, tuples, dictionaries, and sets express different structural rules.
- Control flow translates stated policies into traceable decisions.
- Small functions make analysis reusable, testable, and readable.
- Integration should occur only after each construct can be explained independently.

## Exit ticket

1. Which operation in your problem requires a list rather than a tuple?
2. When would a set communicate the wrong meaning even if the code runs?
3. Why should invalid-data checks appear before ordinary classification branches?
4. Why should a function return a value instead of only printing it?


## References and further reading

- Python Software Foundation. Python 3 documentation: Tutorial and Library Reference.
- VanderPlas, J. (2022). Python Data Science Handbook (2nd ed.).
- CPE15 syllabus, Week 1 course learning plan.

<details>
    <summary><strong>Instructor facilitation note</strong></summary>

            Ask students to predict outputs before execution, compare at least two valid approaches, and explain results in plain language. During live coding, deliberately trigger one common error and model a calm debugging process.

  </details>


In [ ]:

readings = [
    {"station": "ENV-01", "temperature_c": 27.3, "humidity_pct": 65, "pm25_ugm3": 12.4},    # row 1
    {"station": "ENV-02", "temperature_c": 29.8, "humidity_pct": 70, "pm25_ugm3": 18.9},    # row 2
    {"station": "ENV-03", "temperature_c": 31.1, "humidity_pct": 58, "pm25_ugm3": 9.7},     # row 3
    {"station": "ENV-04", "temperature_c": None,  "humidity_pct": 62, "pm25_ugm3": 15.2},   # row 4 (no temperature recorded)
    {"station": "ENV-05", "temperature_c": 28.6, "humidity_pct": 130, "pm25_ugm3": 11.0},   # row 5 (too high humidity recorded)
    {"station": "ENV-06", "temperature_c": 26.9, "humidity_pct": 74, "pm25_ugm3": 22.5},    # row 6
    {"station": "ENV-07", "temperature_c": None, "humidity_pct": 110, "pm25_ugm3": 8.3},    # row 7 (no temp no high humid)
    {"station": "ENV-08", "temperature_c": 27.7, "humidity_pct": 68, "pm25_ugm3": 14.1},    # row 8
]


# validates a single row
# - identifies all existing problems in a room
def validate_env_record(row: dict) -> list:
    # use array to store all errors
    error_messages = []

    humidity = row["humidity_pct"]
    temperature = row["temperature_c"]

    # append error messages as needed

    # none values are not allowed
    if temperature is None:
        error_messages.append("Missing temperature")

    if humidity is None:
        error_messages.append("Missing humidity")

    # check if humid and temp is out of bounds
    humidity_is_out_of_bounds = humidity is not None and not (0 <= humidity <= 100)
    temperature_is_out_of_bounds = temperature is not None and not (-10 <= temperature <= 60)

    if humidity_is_out_of_bounds:
        error_messages.append("Humidity out of bounds (0-100)")

    if temperature_is_out_of_bounds:
        error_messages.append("Temperature out of bounds (-10 to 60)")

    return error_messages


# return the average value as float
# but None for empty list
def get_mean(items: list, amount: int) -> float:
    try:
        return sum(items) / amount
    except ZeroDivisionError:
        return None


# return rounded value if possible
def rounded_value(value):
    return round(value, 2) if value is not None else None;


# validate and include error messages in each row
for row in readings:
    row["errors"] = validate_env_record(row)


# print status of each room
for row in readings:
    status = "usable" if not row["errors"] else ", ".join(row["errors"])
    print(row["station"], "->", status)


# obtain records without errors
row_temp = lambda row: row["temperature_c"]
row_pm25 = lambda row: row["pm25_ugm3"]

usable_rows = list(filter(lambda row: len(row["errors"]) <= 0, readings))
usable_temps = list(map(row_temp, usable_rows))
usable_pm25 = list(map(row_pm25, usable_rows))

print("temps: ", usable_temps)

# calculate total rows, and assert for missing rows
total_records = len(readings)
usable_count = len(usable_rows)
rejected_count = total_records - usable_count

mean_usable_temp = get_mean(usable_temps, usable_count)
mean_usable_pm25 = get_mean(usable_pm25, usable_count)

assert total_records == usable_count + rejected_count


# provide a report dict
report = {
    "total_records": total_records,
    "usable_records": usable_count,
    "rejected_records": rejected_count,
    "mean_usable_temperature_c": rounded_value(mean_usable_temp),
    "mean_usable_pm25_ugm3": rounded_value(mean_usable_pm25),
}

report

ENV-01 -> usable
ENV-02 -> usable
ENV-03 -> usable
ENV-04 -> Missing temperature
ENV-05 -> Humidity out of bounds (0-100)
ENV-06 -> usable
ENV-07 -> Missing temperature, Humidity out of bounds (0-100)
ENV-08 -> usable
temps:  [27.3, 29.8, 31.1, 26.9, 27.7]


{'total_records': 8,
 'usable_records': 5,
 'rejected_records': 3,
 'mean_usable_temperature_c': 28.56,
 'mean_usable_pm25_ugm3': 15.52}

# Interpretation

From the ***8 given recorded values*** of environment sensors ranging from station ENV-01 to ENV-08, there are **three rooms that were rejected**. ENV-04 had missing temperature, ENV-05 had too much himidity, and ENV-07 had missing temperature and high himidity. Overall, the non-rejected rooms had a measured mean temperature of 28.56 degrees celcius and usable pm25 of 15.25